api/src/main/java/org/openmrs/module/oauth2/api/smart/impl/SmartAppLaunchServiceImpl.java

```diff
@@ -0,0 +1,67 @@
+package org.openmrs.module.oauth2.api.smart.impl;
+
+import java.math.BigInteger;
+import java.security.SecureRandom;
+
+import org.openmrs.api.impl.BaseOpenmrsService;
+import org.openmrs.module.oauth2.api.smart.SmartAppLaunchService;
+import org.openmrs.module.oauth2.api.smart.db.hibernate.LaunchValueDAO;
+import org.openmrs.module.oauth2.api.smart.model.LaunchValue;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.transaction.annotation.Transactional;
+
+public class SmartAppLaunchServiceImpl extends BaseOpenmrsService implements SmartAppLaunchService {
+
+	@Autowired
+	LaunchValueDAO launchValueDAO;
+
+	@Transactional
+	@Override
+	public LaunchValue saveOrUpdateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.saveOrUpdate(launchValue);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue updateLaunchValue(LaunchValue launchValue) {
+		return launchValueDAO.update(launchValue);
+	}
+
+	@Transactional(readOnly = true)
+	@Override
+	public LaunchValue getLaunchValue(Integer smartId) {
+		return launchValueDAO.getById(smartId);
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue clearLaunchValue(LaunchValue launchValue) {
+		launchValueDAO.delete(launchValue);
+		return launchValue;
+	}
+
+	@Transactional
+	@Override
+	public LaunchValue merge(LaunchValue launchValue) {
+		return launchValueDAO.merge(launchValue);
+	}
+
+	@Override
+	public void generateAndPersistLaunchValue(LaunchValue launchValue) {
+		generateLaunchValue(launchValue);
+		merge(launchValue);
+		updateLaunchValue(launchValue);
+	}
+
+	private void generateLaunchValue(LaunchValue launchValue) {
+		SecureRandom secureRandom = new SecureRandom();
+		launchValue.setLaunchValue(new BigInteger(130, secureRandom).toString(32));
+	}
+
+	@Override
+	public boolean verifyLaunchValue(LaunchValue launchValue, String receivedLaunchValue) {
```

Comment:
I don't get this method. You are already supplying a launchValue and checking it with a string "receivedValue". Where did the first launchValue come from?

Reply:
In the authorization request.... We will get the client_id from which we can get the smart_id from which we can get the LaunchValue object of the respective SMART app.... Now we can match the launch_value with the recievedLaunchValue...

Reply:
The client_id that you are referring to will be the Client_Identifier (and NOT the client_id that you have joined SmartApp to, they both are different things). Since, smart to launchValue is a 1-1 relation, why not make use of this? Instead of making 4 calls (client_identifier -> client_id -> smart_id -> launchValue) you can verify it in just 2 calls (launchValue -> smart_id) since the rest of the flow (client_identifier -> client_id -> credential_verification) is already done by the module.

Reply:
Anyways, leave this for now. We'll take up optimizations later. I don't want to complicate things for you as of now.

Reply:
Yes.. i got you... But still if we get  launchValue->smartId then... To verify the smart app... We need to check for the smartId->clientId and clientId->clientIdentifier soo it sums up to same.. the either way.... Still yes.. if needed i will change it :)

Reply:
Good work with everything else :)

Reply:
Thanks @mavrk  👍